oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/Sweep2Helper.java

```diff
@@ -0,0 +1,178 @@
+            return -1;
+        }
+    
+        if (status == null || !status.isSweeping()) {
+            // unknown or invalid status, derive from root document if told to check
+            if (checkIfSweep2Necessary && !isSweep2Necessary(ns.getDocumentStore())) {
+                // if no sweep2 is necessary, then mark it so in the settings collection.
+                if (!Sweep2StatusDocument.forceReleaseSweep2LockAndMarkSwept(ns.getDocumentStore(), clusterId)) {
+                    LOG.error("acquireSweep2LockIfNecessary: could not set the sweep2 status. Sweep2 not necessary though.");
+                }
+                // if we concluded that no sweep2 is necessary but failed to update
+                // the status in the settings collection we should continue anyway
+                // (without doing a sweep2 in the background)
+                return -1;
+            } else {
+                return Sweep2StatusDocument.acquireSweep2Lock(ns, clusterId);
+            }
+        }
+        // otherwise the status is "sweeping", which could be by ourselves or by another instance
+    
+        int lockClusterId = status.getLockClusterId();
+        if (lockClusterId == clusterId) {
+            // the local instance was the originator of the sweeping lock, but likely crashed
+            // hence we need to redo the work from scratch as we can't know if we finished it properly
+            LOG.info("acquireSweep2LockIfNecessary : sweep2 status was sweeping, locked by own instance ({}). "
+                    + "Another sweep2 is required.",
+                    clusterId);
+            return status.getLockValue();
+        }
+    
+        // another instance marked as sweeping - check to see if it is still active or it might have crashed
+        ClusterViewDocument clusterViewDoc = ClusterViewDocument.doRead(ns);
+        if (clusterViewDoc == null) {
+            // no clusterView yet - we need to let the caller retry later.
+            // this can happen if the DocumentDiscoveryLiteService wasn't
+            // activated yet or its BackgroundWorker didn't get to
+            // check the view/update it yet.
+            // since that is an async operation, this is a legitimate situation.
+            // but instead of retrying within this method, we leave that to the
+            // caller of this method - hence returning that special value of 0 here.
+            LOG.info("acquireSweep2LockIfNecessary : no clusterView yet, need to determine status later");
+            return 0;
+        }
+    
+        if (clusterViewDoc.getActiveIds().contains(lockClusterId)) {
+            // then another instance is busy sweep2-ing, which is fine.
+            // but we should continue monitoring until that other instance is done
+            LOG.debug("acquireSweep2LockIfNecessary : another instance (id {}) is (still) busy doing a sweep2.",
+                    lockClusterId);
+            return 0;
+        }
```

Comment:
I suggest replacing this with:
```suggestion
        if (ClusterNodeInfoDocument.all(ns.getDocumentStore()).stream()
                .anyMatch(info -> info.getClusterId() == lockClusterId && info.isActive())) {
            // then another instance is busy sweep2-ing, which is fine.
            // but we should continue monitoring until that other instance is done
            LOG.debug("acquireSweep2LockIfNecessary : another instance (id {}) is (still) busy doing a sweep2.",
                    lockClusterId);
            return 0;
        }
```